File: benchmarks/r/10_drug_diagnostics.R
# Benchmark 10: Drug Exposure Diagnostics
# R equivalent: DrugExposureDiagnostics::executeChecks()

source("benchmarks/r/00_helpers.R")
library(DrugExposureDiagnostics)
cat("=== 10: Drug Diagnostics ===\n")
t0 <- proc.time()

con <- dbConnect(duckdb(), dbdir = DB_PATH)
cdm <- cdmFromCon(con = con, cdmSchema = "main", writeSchema = "main", cdmName = "synthea_1k")

# Run diagnostics on clopidogrel (1322184)
result <- executeChecks(
  cdm = cdm,
  ingredients = 1322184,
  checks = c("missing", "exposureDuration", "type", "route", "quantity")
)

# Save summary info about each check
summary_rows <- data.frame(
  check_name = character(),
  n_rows = integer(),
  stringsAsFactors = FALSE
)
for (nm in names(result)) {
  df <- result[[nm]]
  if (!is.null(df) && is.data.frame(df)) {
    summary_rows <- rbind(summary_rows, data.frame(
      check_name = nm,
      n_rows = nrow(df),
      stringsAsFactors = FALSE
    ))
  }
}

save_result(summary_rows, "10_drug_diagnostics")
elapsed <- (proc.time() - t0)[["elapsed"]]
save_timing("10_drug_diagnostics", elapsed)
cdmDisconnect(cdm)
cat("Done in", elapsed, "seconds\n")

File: benchmarks/python/10_drug_diagnostics.py
"""Benchmark 10: Drug Exposure Diagnostics — omopy.drug_diagnostics.execute_checks()"""

import polars as pl
from helpers import Timer, connect_cdm, save_result, save_timing

from omopy.drug_diagnostics import execute_checks

print("=== 10: Drug Diagnostics ===")
t = Timer()
cdm = connect_cdm()

result = execute_checks(
    cdm,
    ingredient_concept_ids=[1322184],  # clopidogrel
    checks=["missing", "exposure_duration", "type", "route", "quantity"],
)

# Combine all check DataFrames
frames = []
for check_name in result:
    df = result[check_name]
    if df is not None and df.height > 0:
        df = df.with_columns(pl.lit(check_name).alias("check_name"))
        frames.append(df)

if frames:
    # All frames may have different schemas, so save individually
    combined = pl.concat(frames, how="diagonal_relaxed")
    save_result(combined, "10_drug_diagnostics")
else:
    save_result(pl.DataFrame({"check_name": [], "note": []}), "10_drug_diagnostics")

save_timing("10_drug_diagnostics", t.elapsed())
print(f"Done in {t.elapsed():.2f} seconds")
